File: go/hackerrank/arrays/array_manipulation.go
package arrays

// Solution to
// https://www.hackerrank.com/challenges/crush/problem
// Too slow
func arrayManipulation(n int32, queries [][]int32) int64 {
	var max int64 = 0

	arr := make([]int64, n, n)

	for _, query := range queries {
		k := query[2]
		for i := query[0] - 1; i < query[1]; i++ {
			arr[i] += int64(k)

			if arr[i] > max {
				max = arr[i]
			}
		}
	}

	return max
}

// Solution to
// https://www.hackerrank.com/challenges/crush/problem
func arrayManipulationWow(n int32, queries [][]int32) int64 {
	var max int64
	arr := make([]int64, n+1)

	for _, q := range queries {
		start := q[0] - 1
		end := q[1]
		k := q[2]

		arr[start] += int64(k)
		arr[end] -= int64(k)
	}

	var tmp int64

	for _, v := range arr {
		tmp += v
		if tmp > max {
			max = tmp
		}
	}

	return max
}

File: javaa/src/main/java/com/example/hackerrank/ArrayManipulation.java
package com.example.hackerrank;

import java.util.List;

public class ArrayManipulation {
  public static long solution(int n, List<List<Integer>> queries) {
    long max = -1;
    int[] arr = new int[n + 2];

    queries.forEach((query) -> {
      arr[query.get(0)] += query.get(2);
      arr[query.get(1) + 1] -= query.get(2);
    });

    long tmp = 0;

    for (int i = 0; i < arr.length; i++) {
      tmp += arr[i];
      max = Math.max(max, tmp);
    }

    return max;
  }

}
